File: app/src/main/java/com/frankhon/fantasymusic/media/AudioPlayerService.java
package com.frankhon.fantasymusic.media;

import static com.frankhon.fantasymusic.utils.Constants.ACTION_NEXT;
import static com.frankhon.fantasymusic.utils.Constants.ACTION_PAUSE;
import static com.frankhon.fantasymusic.utils.Constants.ACTION_PREVIOUS;
import static com.frankhon.fantasymusic.utils.Constants.ACTION_RESUME;
import static com.frankhon.fantasymusic.utils.Constants.ACTION_STOP;
import static com.frankhon.fantasymusic.utils.Notification.sendMediaNotification;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;

import androidx.annotation.Nullable;

import com.frankhon.fantasymusic.IMusicPlayer;
import com.frankhon.fantasymusic.vo.CurrentPlayerInfo;
import com.frankhon.fantasymusic.vo.SimpleSong;
import com.hon.mylogger.MyLogger;

import java.util.List;

/**
 * Created by Frank Hon on 2020/11/1 7:52 PM.
 * E-mail: frank_hon@foxmail.com
 */
public class AudioPlayerService extends Service {

    private final ServiceStub musicPlayer = new ServiceStub();
//    private MediaSessionCompat mediaSessionCompat;

    @Override
    public void onCreate() {
        super.onCreate();
        MyLogger.d("onCreate");
//        mediaSessionCompat = new MediaSessionCompat(this, "MediaService");
//        mediaSessionCompat.setCallback(new MediaSessionCompat.Callback() {
//            @Override
//            public void onStop() {
//                stopForeground(true);
//            }
//        });
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
//        MediaButtonReceiver.handleIntent(mediaSessionCompat, intent);
        String action = intent.getAction();
        MyLogger.d("onStartCommand: " + action);
        if (action != null) {
            switch (action) {
                case ACTION_PREVIOUS:
                    AudioPlayer.previous();
                    break;
                case ACTION_NEXT:
                    AudioPlayer.next();
                    break;
                case ACTION_RESUME:
                    AudioPlayer.resume();
                    break;
                case ACTION_PAUSE:
                    AudioPlayer.pause();
                    break;
                case ACTION_STOP:
                    stopForeground(false);
                    AudioPlayer.release();
                    break;
                default:
                    break;
            }
        }
        return START_REDELIVER_INTENT;
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        MyLogger.d("onBind");
        return musicPlayer;
    }

    @Override
    public boolean onUnbind(Intent intent) {
        MyLogger.d("onUnbind");
//        AudioPlayer.release();
        return super.onUnbind(intent);
    }

    @Override
    public void onDestroy() {
        MyLogger.d("onDestroy");
        super.onDestroy();
    }

    private class ServiceStub extends IMusicPlayer.Stub {

        @Override
        public void play(SimpleSong song) {
            AudioPlayer.play(song);
        }

        @Override
        public void playAndAddIntoPlaylist(SimpleSong song) {
            AudioPlayer.playAndAddIntoPlaylist(song);
            sendMediaNotification(AudioPlayerService.this, AudioPlayer.getCurrentPlayerInfo(), true);
        }

        @Override
        public void addIntoPlaylist(SimpleSong song) {
            boolean isPlaying = AudioPlayer.addIntoPlaylist(song);
            if (isPlaying) {
                sendMediaNotification(AudioPlayerService.this, AudioPlayer.getCurrentPlayerInfo(), true);
            }
        }

        @Override
        public void removeSongFromPlayList(int index) {
            AudioPlayer.removeSongFromPlayList(index);
        }

        @Override
        public void setPlayList(List<SimpleSong> playList, int index) {
            AudioPlayer.setPlaylist(playList, index);
            sendMediaNotification(AudioPlayerService.this, AudioPlayer.getCurrentPlayerInfo(), true);
        }

        @Override
        public void pause() {
            AudioPlayer.pause();
        }

        @Override
        public void resume() {
            AudioPlayer.resume();
        }

        @Override
        public void next() {
            AudioPlayer.next();
        }

        @Override
        public void previous() {
            AudioPlayer.previous();
        }

        @Override
        public void seekTo(int msec) {
            AudioPlayer.seekTo(msec);
        }

        @Override
        public void setPlayMode(String playMode) {
            AudioPlayer.setPlayMode(playMode);
        }

        @Override
        public CurrentPlayerInfo getCurrentPlayerInfo() {
            return AudioPlayer.getCurrentPlayerInfo();
        }
    }
}

File: app/src/main/java/com/frankhon/fantasymusic/media/AudioPlayerService.kt
package com.frankhon.fantasymusic.media

import android.app.Service
import android.bluetooth.BluetoothHeadset
import android.bluetooth.BluetoothProfile
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.Intent.ACTION_HEADSET_PLUG
import android.content.IntentFilter
import android.os.IBinder
import android.support.v4.media.session.MediaSessionCompat
import androidx.media.session.MediaButtonReceiver
import com.frankhon.fantasymusic.IMusicPlayer
import com.frankhon.fantasymusic.data.settings.KEY_NOTIFICATION_STYLE
import com.frankhon.fantasymusic.data.settings.read
import com.frankhon.fantasymusic.media.AudioPlayer.pause
import com.frankhon.fantasymusic.media.AudioPlayer.release
import com.frankhon.fantasymusic.media.AudioPlayer.setPlaylist
import com.frankhon.fantasymusic.media.AudioPlayer.stop
import com.frankhon.fantasymusic.media.notification.MediaButtonCallback
import com.frankhon.fantasymusic.media.notification.sendMediaNotification
import com.frankhon.fantasymusic.utils.ACTION_SCHEDULE_PAUSE_MUSIC
import com.frankhon.fantasymusic.utils.ACTION_STOP
import com.frankhon.fantasymusic.utils.formatTime
import com.frankhon.fantasymusic.vo.CurrentPlayerInfo
import com.frankhon.fantasymusic.vo.SimpleSong
import com.hon.mylogger.MyLogger

/**
 * Created by Frank Hon on 2020/11/1 7:52 PM.
 * E-mail: frank_hon@foxmail.com
 */
class AudioPlayerService : Service() {

    private val musicPlayer = ServiceStub()

    //用来控制通知栏点击事件
    private val mediaSessionCompat by lazy {
        MediaSessionCompat(this, "AudioPlayerService")
            .apply {
                setCallback(MediaButtonCallback())
            }
    }

    // 定时播放完成的回调
    private val pauseMusicReceiver by lazy {
        object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent?) {
                MyLogger.d("pauseMusicReceiver onReceive: ${formatTime(System.currentTimeMillis())}")
                pause()
            }
        }
    }

    // 有线耳机插入和拔出的回调
    private val headsetPlugReceiver by lazy {
        object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent?) {
                val state = intent?.getIntExtra("state", -1) ?: -1
                MyLogger.d("headsetPlugReceiver onReceive: $state")
                if (state == 0) {
                    pause()
                }
            }
        }
    }

    // 蓝牙耳机连接和断开的回调
    private val bluetoothHeadsetReceiver by lazy {
        object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent?) {
                val state = intent?.getIntExtra(
                    BluetoothProfile.EXTRA_STATE,
                    BluetoothProfile.STATE_DISCONNECTED
                ) ?: BluetoothProfile.STATE_DISCONNECTED
                MyLogger.d("bluetoothHeadsetPlugReceiver onReceive: $state")
                if (state == BluetoothProfile.STATE_DISCONNECTED) {
                    pause()
                }
            }
        }
    }

    override fun onCreate() {
        MyLogger.d("onCreate")
        super.onCreate()
        registerReceiver(pauseMusicReceiver, IntentFilter(ACTION_SCHEDULE_PAUSE_MUSIC))
        registerReceiver(headsetPlugReceiver, IntentFilter(ACTION_HEADSET_PLUG))
        registerReceiver(
            bluetoothHeadsetReceiver,
            IntentFilter().apply {
                addAction(BluetoothHeadset.ACTION_CONNECTION_STATE_CHANGED)
                addAction(BluetoothHeadset.ACTION_AUDIO_STATE_CHANGED)
            }
        )
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        MediaButtonReceiver.handleIntent(mediaSessionCompat, intent)
        val action = intent.action
        MyLogger.d("onStartCommand: $action")
        if (action == ACTION_STOP) {
            stopMusic()
        }
        return START_REDELIVER_INTENT
    }

    override fun onBind(intent: Intent): IBinder {
        MyLogger.d("onBind")
        return musicPlayer
    }

    override fun onUnbind(intent: Intent): Boolean {
        MyLogger.d("onUnbind")
        // AudioPlayer.release();
        return super.onUnbind(intent)
    }

    override fun onDestroy() {
        MyLogger.d("onDestroy")
        super.onDestroy()
        unregisterReceiver(pauseMusicReceiver)
        unregisterReceiver(headsetPlugReceiver)
        unregisterReceiver(bluetoothHeadsetReceiver)
        release()
    }

    private fun stopMusic() {
        stopForeground(false)
        stop()
        release()
    }

    private fun sendMediaNotification() {
        read {
            sendMediaNotification(
                it[KEY_NOTIFICATION_STYLE] ?: true,
                this@AudioPlayerService,
                AudioPlayer.getCurrentPlayerInfo(),
            )
        }
    }

    private inner class ServiceStub : IMusicPlayer.Stub() {
        override fun play(song: SimpleSong) {
            AudioPlayer.play(song)
            sendMediaNotification()
        }

        override fun playAndAddIntoPlaylist(song: SimpleSong) {
            AudioPlayer.playAndAddIntoPlaylist(song)
            sendMediaNotification()
        }

        override fun addIntoPlaylist(song: SimpleSong) {
            val isPlaying = AudioPlayer.addIntoPlaylist(song)
            if (isPlaying) {
                sendMediaNotification()
            }
        }

        override fun removeSongFromPlayList(index: Int) {
            val isPlaylistEmpty = AudioPlayer.removeSongFromPlayList(index)
            if (isPlaylistEmpty) {
                stopMusic()
            }
        }

        override fun setPlayList(playList: List<SimpleSong>, index: Int) {
            setPlaylist(playList, index)
            sendMediaNotification()
        }

        override fun pause() {
            AudioPlayer.pause()
        }

        override fun resume() {
            AudioPlayer.resume()
        }

        override fun next() {
            AudioPlayer.next()
        }

        override fun previous() {
            AudioPlayer.previous()
        }

        override fun seekTo(msec: Int) {
            AudioPlayer.seekTo(msec)
        }

        override fun setPlayMode(playMode: String) {
            AudioPlayer.setPlayMode(playMode)
        }

        override fun getCurrentPlayerInfo(): CurrentPlayerInfo {
            return AudioPlayer.getCurrentPlayerInfo()
        }
    }
}